moveit_experimental/jog_arm/include/jog_arm/jog_arm_server.h

```diff
@@ -0,0 +1,228 @@
+
+// For collision checking thread
+void* collisionCheck(void* threadid);
+
+// Shared variables
+geometry_msgs::TwistStamped g_command_deltas;
+pthread_mutex_t g_command_deltas_mutex;
+
+sensor_msgs::JointState g_joints;
+pthread_mutex_t g_joints_mutex;
+
+trajectory_msgs::JointTrajectory g_new_traj;
+pthread_mutex_t g_new_traj_mutex;
+
+bool g_imminent_collision(false);
+pthread_mutex_t g_imminent_collision_mutex;
+
+bool g_zero_trajectory_flag(false);
+pthread_mutex_t g_zero_trajectory_flagmutex;
+
+// ROS subscriber callbacks
+void deltaCmdCB(const geometry_msgs::TwistStampedConstPtr& msg);
+void jointsCB(const sensor_msgs::JointStateConstPtr& msg);
+
+int readParameters(ros::NodeHandle& n);
+
+// ROS params to be read
+struct jog_arm_parameters
+{
+  std::string move_group_name, joint_topic, command_in_topic, command_frame, command_out_topic, planning_frame,
+      warning_topic;
+  double linear_scale, rotational_scale, singularity_threshold, hard_stop_singularity_threshold, low_pass_filter_coeff,
+      publish_period, incoming_command_timeout;
+  bool gazebo, collision_check;
+} g_parameters;
+
+/**
+ * Class LowPassFilter - Filter the joint velocities to avoid jerky motion.
+ */
+class LowPassFilter
+{
+public:
+  LowPassFilter(double low_pass_filter_coeff);
+  double filter(const double& new_msrmt);
+  void reset(double data);
+  double filter_coeff_ = 10.;
+
+private:
+  double prev_msrmts_[3] = { 0., 0., 0. };
+  double prev_filtered_msrmts_[2] = { 0., 0. };
+};
+
+LowPassFilter::LowPassFilter(double low_pass_filter_coeff)
+{
+  filter_coeff_ = low_pass_filter_coeff;
+}
+
+void LowPassFilter::reset(double data)
+{
+  prev_msrmts_[0] = data;
+  prev_msrmts_[1] = data;
+  prev_msrmts_[2] = data;
+
+  prev_filtered_msrmts_[0] = data;
+  prev_filtered_msrmts_[1] = data;
+}
+
+double LowPassFilter::filter(const double& new_msrmt)
+{
+  // Push in the new measurement
+  prev_msrmts_[2] = prev_msrmts_[1];
+  prev_msrmts_[1] = prev_msrmts_[0];
+  prev_msrmts_[0] = new_msrmt;
+
+  double new_filtered_msrmt = (1 / (1 + filter_coeff_ * filter_coeff_ + 1.414 * filter_coeff_)) *
+                              (prev_msrmts_[2] + 2 * prev_msrmts_[1] + prev_msrmts_[0] -
+                               (filter_coeff_ * filter_coeff_ - 1.414 * filter_coeff_ + 1) * prev_filtered_msrmts_[1] -
+                               (-2 * filter_coeff_ * filter_coeff_ + 2) * prev_filtered_msrmts_[0]);
+
+  // Store the new filtered measurement
+  prev_filtered_msrmts_[1] = prev_filtered_msrmts_[0];
+  prev_filtered_msrmts_[0] = new_filtered_msrmt;
+
+  return new_filtered_msrmt;
+}
+
+/**
+ * Class JogCalcs - Perform the Jacobian calculations.
+ */
+class JogCalcs
+{
+public:
+  JogCalcs(const std::string& move_group_name);
+
+protected:
+  ros::NodeHandle nh_;
+
+  moveit::planning_interface::MoveGroupInterface arm_;
+
+  geometry_msgs::TwistStamped cmd_deltas_;
+
+  sensor_msgs::JointState incoming_jts_;
+
+  typedef Eigen::Matrix<double, 6, 1> Vector6d;
```

Comment:
this typedef is already availble in eigen https://eigen.tuxfamily.org/dox/group__matrixtypedefs.html#gaabb0b4639dc0b48e691e02e95873b0f1